app/src/main/java/io/appium/uiautomator2/model/internal/GestureController.java

```diff
@@ -60,31 +55,113 @@ private static Method extractPerformGestureMethod(Object wrappedInstance) {
+        return getDevice().performActionAndWait(new GestureRunnable(gestures), condition, timeout);
     }
 
     public void click(Point point) {
-        // new PointerGesture(point).pause(0);
-        Object gesture;
-        try {
-            gesture = getPointerGestureConstructor().newInstance(point);
-        } catch (IllegalAccessException | InstantiationException | InvocationTargetException e) {
-            throw new IllegalStateException(String.format("Cannot perform click gesture at %s", point), e);
+        performGesture(new PointerGesture(point).pause(0L));
+    }
+
+    public void longClick(Point point, @Nullable Long durationMs) {
+        performGesture(new PointerGesture(point)
+                .pause(durationMs == null ? ViewConfiguration.getLongPressTimeout() : durationMs));
+    }
+
+    private static int checkSpeed(int speed) {
+        if (speed < 0) {
+            throw new IllegalArgumentException("Speed cannot be negative");
         }
-        Method pauseMethod = method(getPointerGestureClass(), "pause", long.class);
-        gesture = invoke(pauseMethod, gesture, 0L);
-        performGesture(gesture);
+        return speed;
+    }
+
+    private static float checkPercent(float percent) {
+        if (percent < 0.0f || percent > 1.0f) {
+            throw new IllegalArgumentException("Percent must be between 0.0f and 1.0f");
+        }
+        return percent;
+    }
+
+    public void drag(Point start, Point end, @Nullable Integer speed) {
+        int dragSpeed = speed == null ? Gestures.getDefaultDragSpeed() : checkSpeed(speed);
+        performGesture(gestures.drag(start, end, dragSpeed));
+    }
+
+    public void pinchClose(Rect area, float percent, @Nullable Integer speed) {
+        int pinchSpeed = speed == null ? Gestures.getDefaultPinchSpeed() : checkSpeed(speed);
+        performGesture(gestures.pinchClose(area, checkPercent(percent), pinchSpeed));
+    }
+
+    public void pinchOpen(Rect area, float percent, @Nullable Integer speed) {
+        int pinchSpeed = speed == null ? Gestures.getDefaultPinchSpeed() : checkSpeed(speed);
+        performGesture(gestures.pinchOpen(area, checkPercent(percent), pinchSpeed));
+    }
+
+    public void swipe(Rect area, Direction direction, float percent, @Nullable Integer speed) {
+        int swipeSpeed = speed == null ? Gestures.getDefaultSwipeSpeed() : checkSpeed(speed);
+        performGesture(gestures.swipe(area, direction, checkPercent(percent), swipeSpeed));
+    }
+
+    public boolean scroll(Rect area, Direction direction, float percent, @Nullable Integer speed) {
+        if (percent < 0.0f) {
+            throw new IllegalArgumentException("Percent must be greater than 0.0f");
+        }
+
+        Direction swipeDirection = Direction.reverse(direction);
+        int scrollSpeed = speed == null ? Gestures.getDefaultScrollSpeed() : checkSpeed(speed);
+        for (float swipePercent = percent; swipePercent > 0.0f; swipePercent -= 1.0f) {
```

Comment:
Codacy found an issue: [Dont use floating point for loop indices. If you must use floating point, use double.](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6222243)